### ISSUE TITLE
Reconnect automatically

### ISSUE THREAD

**Issue by mleonhard** [featurehub-io/featurehub-flutter-sdk#17]
I wish to use `featurehub_sse_client` in an iOS/Android Flutter app.  I need the library to automatically reconnect on error.  It should handle all kinds of errors: DNS lookup failure, TCP connect failure, connection reset, TLS negotiation failure, HTTP 4xx & 5xx responses, protocol errors, and clean connection shutdown.  I need it to reconnect immediately once and then back off to once every 10 seconds or so.

Does the library have this reconnect behavior?

If so, is there a plan to add tests for it?

**Comment by rvowles** [comment]
Hi @mleonhard 

It doesn't at the moment as it assumes you "got it wrong", so there is no point in trying again. I don't see any reason why we couldn't add these capabilities?

**Comment by schieberegister** [comment]
Hey guys,

I'm also trying to implement reconnects. My first approach was the following code. Unfortunately  the event source will always report ReadyState as OPEN event if i shutdown the server.

```dart
Future<void> events() async {
  try {
    EventSource eventSource =
        await EventSource.connect("http://localhost:8003/events");
    eventSource.listen((event) {
      print("New event:");
      print("  event: ${event.event}");
      print("  data: ${event.data}");
    });
    while (true) {
      print(eventSource.client);
    }
  } catch (e) {
    print(e);
  }
}
```

**Comment by rvowles** [comment]
I'm happy to receive PRs for this, but EventSourcing has traditionally been used in our SDK only for the Web clients - and there it is underpinned by the native browser eventsource (which does keep reconnecting). We do not recommend it for any clients for mobile phones as it is  a huge battery drain to keep a TCP connection open and generally not a recommended practice. 

With the advent of native in Mac/Windows/Linux, having this capability would seem logical, but it would be in our backlog for the time being. So PRs to resolve it for native would be welcome.

**Comment by schieberegister** [comment]
I see. Because i‘m running on flutter-pi, most assumptions for classical devices do not apply. That is why I implemeted my own little sse client based on universal_html‘s event source.

A compromise for people using your library would be, to expose a function or stream like readyState, reporting the state of the underlying client/socket so they could trigger the reconnect e.g. in an async function.

**Comment by rvowles** [comment]
There is a PR associated now that does just that @schieberegister

**Comment by schieberegister** [comment]
@rvowles This is great news. Thank you for your effort.

**Comment by rvowles** [comment]
the ability to do this was released as 1.3.2 - you can pass in your own streamcontroller (I use a broadcast stream) and key updates) and there is now a retry() method which you can call if you get a CLOSED event.
